tidy3d/components/parameter_perturbation.py

```diff
@@ -1101,3 +1116,239 @@ def is_complex(self) -> bool:
         """Whether perturbation is complex valued."""
 
         return np.any([p.is_complex for p in self.perturbation_list])
+
+
+class PermittivityPerturbation(Tidy3dBaseModel):
+    """A general medium perturbation model which is defined through perturbation to
+    permittivity and conductivity.
+
+    Example
+    -------
+    >>> from tidy3d import LinearChargePerturbation, LinearHeatPerturbation, PermittivityPerturbation, C_0
+    >>>
+    >>> heat_perturb = LinearHeatPerturbation(
+    ...     temperature_ref=300,
+    ...     coeff=0.001,
+    ... )
+    >>> charge_perturb = LinearChargePerturbation(
+    ...     electron_ref=0,
+    ...     electron_coeff=0.0001,
+    ...     hole_ref=0,
+    ...     hole_coeff=0.0002,
+    ... )
+    >>> deps = ParameterPerturbation(heat=heat_perturb)
+    >>> dsigma = ParameterPerturbation(charge=charge_perturb)
+    >>> permittivity_pb = PermittivityPerturbation(deps=deps, dsigma=dsigma)
+    """
+
+    deps: Optional[ParameterPerturbation] = pd.Field(
```

Comment:
Maybe `delta_eps`, etc. is better than just `d`. I forget if we have any convention about this anywhere already. We use `dn_dwvl` [here](https://github.com/flexcompute/tidy3d/blob/7f69690abd87b9cf6e76f5b5d0e341ad7b47ba05/tidy3d/components/medium.py#L3065-L3089) but this is an actual derivative..

Reply:
yeah, I wasn't sure about this naming too. When we refer to refraction index we typically just use `n` and `k`, so having `dn` and `dk` kind of made sense, short and clear. And then I just extended it to eps and sigma. However, for permittivity and conductivity we use actual `permittivity` and `conductivity` as class field names. That is why when perturbation classes were introduced initially, I followed the convention of adding `_perturbation`, that is, `permittivity_perturbation`, `conductivity_perturbation`, `eps_inf_perturbation`, `poles_perturbation`. Also because things like `dpoles` or `delta_poles` seemed worse.

Having said that, different options I see to use in `PermittivityPerturbation` and `IndexPerturbation`:
- Stick with the `_perturbation` convection. That is, use either 
`n_perturbation`, `k_perturbation`, `permittivity_perturbation`, `conductivity_perturbation`
or `index_perturbation`, `extinction_perturbation`, `permittivity_perturbation`, `conductivity_perturbation`. In this case, defining perturbations directly and through `perturbation_spec` would be uniform, but the downside it's quite a mouthful:
```
med_direct = td.PerturbationMedium(
    permittivity=5,
    permittivity_perturbation=td.ParameterPerturbation(...),
)

med_spec_equivalent = td.PerturbationMedium(
    permittivity=5,
    perturbation_spec=td.PermittivityPerturbation(
        permittivity_perturbation=td.ParameterPerturbation(...),
    ),
)

med_spec_via_n = td.PerturbationMedium(
    permittivity=5,
    perturbation_spec=td.IndexPerturbation(
        n_perturbation=td.ParameterPerturbation(...),
    ),
)
```

- Use shortened field names similar to what I have now `dn`, `dk`, `deps`, `dsigma`, but something more explicit like `delta_n`, `delta_k`, `delta_eps`, `delta_sigma` or, on the contrary, just plain `n`, `k`, `eps`, `sigma` since they will be inside `SomethingPerturbation` already. This somewhat breaks uniformity, but maybe it can be considered as highlighting the difference between defining perturbation models directly on medium fields or as `perturbation_spec`
```
med_direct = td.PerturbationMedium(
    permittivity=5,
    permittivity_perturbation=td.ParameterPerturbation(...),
)

med_spec_equivalent = td.PerturbationMedium(
    permittivity=5,
    perturbation_spec=td.PermittivityPerturbation(
        delta_eps=td.ParameterPerturbation(...),
    ),
)

med_spec_via_n = td.PerturbationMedium(
    permittivity=5,
    perturbation_spec=td.IndexPerturbation(
        delta_n=td.ParameterPerturbation(...),
    ),
)

Reply:
also, open to discuss naming of new specs as well. For example:
`IndexPerturbation` -> `NKPerturbation`
`PermittivityPerturbation` -> `DieletricPerturbation`

Reply:
I think I prefer the original `IndexPerturbation`, `PermittivityPerturbation`

Reply:
Agree re IndexPerturbation, PermittivityPerturbation.

About the rest, I am not sure. `_perturbation` has its merits but is indeed quite verbose. Regarding your comment that it could just be `n` instead of `dn` or `delta_n`, could it also just be `delta` since the `SomethingPerturbation` defines what exactly is being perturbed?

```
med_spec_equivalent = td.PerturbationMedium(
    permittivity=5,
    perturbation_spec=td.PermittivityPerturbation(
        delta=td.ParameterPerturbation(...),
    ),
) 
```

If not then I do prefer `delta_n`, etc. to `dn`. Or at least `d_n`. Otherwise you get awkward to parse kwarg names like `deps` or `dpoles`.

Reply:
yeah, just both `IndexPerturbation` and `PermittivityPerturbation` represent perturbation of two parameters each, n and k for `IndexPerturbation`, eps and sigma for `PermittivityPerturbation`. Like this:
```
med_1 = td.PerturbationMedium(
    permittivity=5,
    perturbation_spec=td.PermittivityPerturbation(
        delta_eps=td.ParameterPerturbation(...),
        delta_sigma=td.ParameterPerturbation(...),
    ),
) 

med_2 = td.PerturbationMedium(
    permittivity=5,
    perturbation_spec=td.IndexPerturbation(
        delta_n=td.ParameterPerturbation(...),
        delta_k=td.ParameterPerturbation(...),
        freq=freq0.
    ),
) 
```

Reply:
btw, any preference between `freq` and `frequency`? It looks like we use both options in our code, for example, `eps_model(frequency=...)` but `from_nk(n=..., k=..., freq=...)`

Reply:
we unfortunately use `f` in the DataArrays 😢 . (and also mixed notation `freq0` and `f_width` in the gaussian pulse... ) what do you think? I'm maybe inclined towards `freq` to be consistent with `from_nk`